### PR TITLE
Deferred blobs

### PR DESCRIPTION
**Pull request by millerdev**
Add `DeferredBlobMixin` and `BlobHelper`, which will be used when `XFormInstance` attachments are migrated to the blob db.

Review commits individually.

@dannyroberts cc @snopoke @esoergel

### CODE REVIEW

File: corehq/blobs/tests/test_mixin.py [hunk 1]
```diff
@@ -90,6 +90,50 @@ def test_document_blobs(self):
...
+        content = b"<xml />"
+        obj.deferred_put_attachment(content, name, content_type="text/xml")
+        self.assertEqual(obj.fetch_attachment(name), content)
+
+    def test_delete_attachment_deletes_unsaved_blob(self):
+        obj = self.make_doc(DeferredPutBlobDocument)
+        name = "test.1"
+        content = b"<xml />"
+        obj.deferred_put_attachment(content, name, content_type="text/xml")
+        self.assertTrue(obj.delete_attachment(name))
```
Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: For completeness should also check that `fetch_attachment` returns nothing

File: corehq/blobs/tests/test_mixin.py [hunk 2]
```diff
@@ -90,6 +90,50 @@ def test_document_blobs(self):
...
+
+    def test_save_persists_unsaved_blob(self):
+        obj = self.make_doc(DeferredPutBlobDocument)
+        name = "test.1"
+        content = StringIO(b"content")
+        obj.deferred_put_attachment(content, name, content_type="text/plain")
+        obj.save()
+        self.assertTrue(obj.saved)
```
Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: Could be better to re-fetch the object from the DB before validating it

File: corehq/blobs/tests/test_mixin.py [hunk 3]
```diff
@@ -463,6 +466,239 @@ def setUpClass(cls):
...
+            raise Exception("fail!")
+        couch = FakeCouchDatabase()
+        couch.fetch_attachment = fetch_fail
+        obj = mod.BlobHelper({
+            "_id": "fetch-fail",
+            "external_blobs": {"not-found.txt": {"id": "hahaha"}},
+        }, couch)
+        self.assertFalse(obj.migrating_blobs_from_couch)
+        with self.assertRaises(mod.ResourceNotFound):
+            obj.fetch_attachment("not-found.txt")
```
Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: Not really following this test. I would have thought that it shoudn't raise an exception here

Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: Re-checked the code and the test is correct - just the test name is misleading

Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: Never mind - finally got my head around this

File: corehq/blobs/tests/test_mixin.py (hunk 2 above)
Comment (millerdev) [inline corehq/blobs/tests/test_mixin.py]: All couch db is mocked in these tests to keep them fast. This line validates that the base [`obj.save()`](https://github.com/dimagi/commcare-hq/pull/11523/files#diff-8a2de1f87337170d35b54199a1fc7b1eR724) was called, which is assumed to actually save the document in a non-test environment.

Comment (millerdev) [inline corehq/blobs/tests/test_mixin.py]: Does that explanation make sense? Did I understand your suggestion correctly?

File: corehq/blobs/tests/test_mixin.py (hunk 1 above)
Comment (millerdev) [inline corehq/blobs/tests/test_mixin.py]: https://github.com/dimagi/commcare-hq/pull/11523/commits/f43ab4ba98586eb777f1e6bb2bef20cb01272b17

File: corehq/blobs/tests/test_mixin.py (hunk 2 above)
Comment (snopoke) [inline corehq/blobs/tests/test_mixin.py]: :+1:

**Comment by snopoke** [comment]
:+1: from me but will let @dannyroberts take a look

**Comment by dannyroberts** [comment]
Looks good to me too.

**Pull request merged**